jitstats.lua

```diff
@@ -0,0 +1,462 @@
+        end)
+    end
+
+    if curr_stitched then
+        stats.stitch_aborts = stats.stitch_aborts + 1
+    elseif curr_parentid then
+        debugprint("ABORT(%d): parent = %d, %s", tr, curr_parentid, reason)
+        stats.side_aborts = stats.side_aborts + 1
+    else
+        debugprint("ABORT(%d): %s", tr, reason)
+    end
+    
+    local bc = funcbc(start_func, start_pc)
+    local bcname = bc and bcnames[band(bc, 0xff)]
+    if bcname == "IFUNCF" then
+        print"blacklist"
+        stats.func_blacklisted = stats.func_blacklisted + 1
+    elseif bcname == "ILOOP" or bcname == "IFORL" or bcname == "IITERL" then
+        stats.loop_blacklisted = stats.loop_blacklisted + 1
+    end
+
+    -- We sometimes don't can't find bc
+    bc = funcbc(func, pc)
+    local opcode = bc and band(bc, 0xff)
+    
+    if code == TRERR_LLEAVE then
+        debugprint("LLEAVE: %d/%s", opcode, bcnames[opcode])
+    elseif code == TRERR_LUNROLL then
+        
+    elseif code == TRERR_NYIBC then
+        bcname = bcnames[errinfo]
+        stats.nyi_bc[bcname] = (stats.nyi_bc[bcname] or 0) + 1
+    end
+
+    -- Use internal enum name of the trace abort error if possible
+    local tracerror = traceerr_fromid[code] or vmdef.traceerr[code]
+    stats.abort_counts[tracerror] = (stats.abort_counts[tracerror] or 0) + 1
+end
+
+function vmevent.flush()
+    debugprint("Full trace flush")
+    stats.traceflush = stats.traceflush + 1
+end
+
+function vmevent.texit(id, exitno)
+    debugprint("EXIT(%d): exit = %d", id, exitno)
+    stats.exits = stats.exits + 1
+end
+
+local function vmeventcb(what, ...)
+    local cb = vmevent[what]
+    if cb then 
+        cb(...)
+    end
+end
+
+function jitstats.start()
+    if jitstats.isrunning then
+        return false
+    end
+
+    jit.attach(vmeventcb, "trace")
+    jit.attach(vmevent.texit, "texit")
+    jitstats.isrunning = true
+    return true
+end
+
+function jitstats.stop()
+    if not jitstats.isrunning then
+        return false
+    end
+
+    jit.attach(vmeventcb)
+    jit.attach(vmevent.texit)
+    jitstats.isrunning = false
+    return true
+end
+
+local last_starts, last_exits
+
+function jitstats.reset(statstbl)
+    statstbl = statstbl or stats
+    statstbl.starts = 0
+    statstbl.aborts = 0
+    statstbl.abort_counts = {}
+    statstbl.completed = 0
+    statstbl.linktypes = {}
+    statstbl.stitch_starts = 0
+    statstbl.stitch_aborts = 0
+    statstbl.stitch_completed = 0
+    statstbl.side_starts = 0
+    statstbl.side_aborts = 0  
+    statstbl.side_completed = 0
+    statstbl.nyi_bc = {}
+    statstbl.exits = 0
+    statstbl.traceflush = 0
+    statstbl.func_blacklisted = 0
+    statstbl.loop_blacklisted = 0
+end
+
+local function table_copy(src, dst)
+    dst = dst or {}
+    for k,v in pairs(src) do
+        if type(v) ~= "table" then
+            dst[k] = v
+        else
+            dst[k] = table_copy(v, dst[k])
+        end
+    end
+    return dst
+end
```

Comment:
- maybe a warning about not being safe with self-referenting tables?
- on full `if...then...else...end` constructs, I typically find more readable to put the "positive" condition in as `then`, and the negative in the `else`.  like:
```
if type(v) == 'table' then
   dst[k] = table_copy(v, dst[k])
else
   dst[k] = v
end
```
- what about this form? (untested!):
```
local function table_copy(src, dst)
    if type(src) ~= 'table' then return src end
    dst = dst or {}
    for k, v in pairs(src) do
        dst[k] = table_copy(v, dst[k])
    end
    return dst
end
```

Reply:
I just went with the most common one first.